juntagrico/queryset/subscription.py

```diff
@@ -0,0 +1,155 @@
+            # convert trial days into duration. Minus 1 to end up at the end of the last trial day, e.g., 1. + 30 days = 30. (not 31.)
+            parts__type__trial_duration=ExpressionWrapper(F('parts__type__trial_days') * self.one_day, DurationField()) - self.one_day,
+            # find (assumed) deactivation date of part
+            parts__forecast_final_date=Least(
+                end,  # limit final date to period of interest
+                Case(
+                    # use deactivation date if set
+                    When(parts__deactivation_date__isnull=False,
+                         then='parts__deactivation_date'),
+                    # on trial subs without deactivation date assume they will last for trial duration.
+                    When(parts__type__trial_days__gt=0,
+                         then=Cast(F('parts__activation_date') + F('parts__type__trial_duration'), DateField())),
+                    # otherwise default to end of period
+                    default=end,
+                    output_field=DateField()
+                )
+            ),
+            # number of days subscription part is actually active within period of interest. Add a day because activation day should also count to duration
+            parts__duration_in_period=F('parts__forecast_final_date') - Greatest('parts__activation_date', start) + self.one_day,
+            parts__duration_in_period_float=Greatest(
+                0.0,  # ignore values <0 resulting from parts outside the period of interest
+                # Tested on postgres (ExtractDay) and SQLite (Cast)
+                ExtractDay('parts__duration_in_period')
+                if connection.features.has_native_duration_field else
+                Cast('parts__duration_in_period', FloatField()) / self.microseconds_in_day,
+                output_field=FloatField()
+            ),
+            # number of days within which the assignments are required
+            parts__reference_duration=Case(
+                When(parts__type__trial_days__gt=0,
+                     then='parts__type__trial_days'),
+                default=self.days_in_year,
+                output_field=FloatField()
+            ),
+            parts__required_assignments_discount=Case(
+                # ignore parts that have not startet yet
+                When(parts__activation_date__isnull=True,
+                     then=0.0),
+                # get discount ratio for required assignments
+                default=F('parts__duration_in_period_float') / F('parts__reference_duration')
+            )
+        ).annotate(  # annotate the final results
+            required_assignments=Round(Sum(F('parts__type__required_assignments') * F('parts__required_assignments_discount'), default=0.0)),
+            required_core_assignments=Round(Sum(F('parts__type__required_core_assignments') * F('parts__required_assignments_discount'), default=0.0)),
+        )
+
+    def annotate_assignments_progress(self, prefix=''):
```

Comment:
I agree that this may not be faster or better on memory, but also not worse. But the alternative would be to either calculate it in all places where it is needed or have a method on the subs to do that, but it would also require that the annotations were performed before.
Also blocking the table should not be an issue, as django is blocked anyways during the request and the tables won't have other things to do I guess.

Maybe I can remove the prerequisites of this method by just calling the necessary other annotations inside of it. I'll look into this.

Reply:
`annotate_assignments_progress` now works on it's own. An exception is raised on unsupported usage (e.g. annotating two date ranges on the same qs).